### PR TITLE
Remove read pool

### PR DESCRIPTION
**Pull request by breeswish**
## What have you changed? (mandatory)

This PR removes the read pool, so that coprocessor / storage component will depend *less* on the server component. According to the diff, the code deletion is more than code addition.

## What are the type of the changes? (mandatory)

- Engineering (engineering change which doesn't change any feature or fix any issue)

## How has this PR been tested? (mandatory)

Existing tests are modified.

### CODE REVIEW

**Comment by siddontang** [comment]
Great work

How about the compiling time?

File: components/test_coprocessor/src/fixture.rs [hunk 1]
```diff
@@ -5,6 +5,7 @@ use super::*;
 use kvproto::kvrpcpb::Context;
 
 use tidb_query::codec::Datum;
+use tikv::config::CoprocessorReadPoolConfig;
```
Comment (siddontang) [inline components/test_coprocessor/src/fixture.rs]: CoprocessorReadPoolConfig is too long, can you shorten it?

File: components/tikv_util/src/future_pool/builder.rs [hunk 2]
```diff
@@ -76,4 +111,14 @@ impl Builder {
         let pool = Arc::new(self.inner_builder.build());
         super::FuturePool { pool, env }
     }
+
+    /// Builds the FuturePool which limits max tasks.
+    pub fn build_with_task_limit(&mut self) -> super::TaskLimitedFuturePool {
```
Comment (siddontang) [inline components/tikv_util/src/future_pool/builder.rs]: why not let the FuturePool support limiting task directly? I think we can check task limited in FuturePool spawn directly.

Comment (breeswish) [reply components/tikv_util/src/future_pool/builder.rs]: I marked a TODO for it. This looks to be pretty easy for contributors to change. Currently `FuturePool` is also used in somewhere, which provides different signatures compared to `TaskLimitedFuturePool`. More specifically, `FuturePool` provides non-failure interfaces while `TaskLimitedFuturePool` provides interfaces wrapped by `Result`.

Comment (siddontang) [reply components/tikv_util/src/future_pool/builder.rs]: I prefer you providing FuturePool at first. 

Which mean first supporting task limited for the FuturePool (you may need to handle result of spawn too), then remove the ReadPool.

Comment (siddontang) [reply components/tikv_util/src/future_pool/builder.rs]: /cc @BusyJay

Comment (BusyJay) [reply components/tikv_util/src/future_pool/builder.rs]: Supporting limited task first makes sense to me.

**Comment by breeswish** [comment]
@siddontang Introduced Full error for all FuturePools. By default there is no limit (i.e. safe to `unwrap()` ).

About compiling time, I don't think there could be any differences since it does not change crate relations and only modifies a few LOC.

**Comment by siddontang** [comment]
can you split your PR into two?

one is to support the task limited future pool, the other is to remove readpool? 

@breeswish

**Comment by breeswish** [comment]
@siddontang The change of supporting task limited future pool is very few, only 20 lines of code (and is copied from the previous ReadPool): https://github.com/tikv/tikv/pull/5220/files#diff-7aec89bf97143ed35d6305fb3fda14e5R58  I think the separation may not help review much.

**Comment by siddontang** [comment]
@breeswish 

I still insist on that we should do one thing in one PR, IMO, there are two things in your PR.

**Comment by breeswish** [comment]
@siddontang Extracted to https://github.com/tikv/tikv/pull/5301 now.

**Comment by AndreMouche** [comment]
Please fix the conflicts.

File: src/config.rs [hunk 3]
```diff
@@ -1171,23 +1200,23 @@ impl Default for StorageReadPoolConfig {
...
+            stack_size: ReadableSize::mb(DEFAULT_READPOOL_STACK_SIZE_MB),
         }
     }
 }
 
 const DEFAULT_COPROCESSOR_READPOOL_CONCURRENCY: usize = 8;
 
 readpool_config!(
-    CoprocessorReadPoolConfig,
+    CoprReadPoolConfig,
```
Comment (youjiali1995) [inline src/config.rs]: Copr or Cop?

Comment (breeswish) [reply src/config.rs]: Both are ok. I prefer Copr now because "cop" is a valid word..XD

File: src/server/node.rs [hunk 4]
```diff
@@ -36,7 +36,7 @@ const CHECK_CLUSTER_BOOTSTRAPPED_RETRY_SECONDS: u64 = 3;
 pub fn create_raft_storage<S>(
     engine: RaftKv<S>,
     cfg: &StorageConfig,
-    read_pool: ReadPool,
+    read_pools: Vec<FuturePool>,
```
Comment (youjiali1995) [inline src/server/node.rs]: read_pool or future_pools?

Comment (breeswish) [reply src/server/node.rs]: it is a pool for reading, implemented using FuturePool. Both names are fine.

**Pull request merged**